further refine incremental result types

yaacovCR opened the pull request:
continue work with #4313

Fixes the types for Subsequent Execution Results. Currently, the generic argument for subsequent results is passed to deferred and streamed results. By providing two different generic arguments, we can get properly types individual members of the incremental array.

yaacovCR committed:
further refine incremental result types

continue work with #4313

JoviDeCroock commented on src/execution/types.ts:
```diff
@@ -36,12 +36,13 @@ export interface FormattedExecutionResult<
 export interface ExperimentalIncrementalExecutionResults<
   TInitial = ObjMap<unknown>,
-  TSubsequent = unknown,
+  TData = ObjMap<unknown>,
+  TItems = ReadonlyArray<unknown>,
```
Not sure what we are trying to achieve here but should we be explicit about what these types can be?

i.e. if Data is always going to be an object should we use `extends Record<string, unknown>`

Furthermore `TItems` sounds like it's always going to be an array so why don't we make this a `TIncrementalItem` and wrap it in the `ReadonlyArray` ourselves? Is the purpose of this wrapping so that folks can avoid passing in a readonly one? If so then we still need to push people into the happy path by constraining the inputs.

yaacovCR replied:
Hmmm, basically, I ran into type conflicts when working on the transformer in #4319 because TData was mapped to TItems, and this was a simple fix for that.

Taking a step back, I think we should honestly at this point remove all of the generics from the incremental results @robrichard . As far as I understand, these are used by tools like GraphQL CodeGen to provide a typed experience based on the query, but it is possible/likely that typed access to the incremental data will be from the merged object anyway, after all the incremental bits have been merged in, and we don't need to provide these generics at all.

I think we may have simply started off supplying them because the IncrementalDeferResult is modelled after an ExecutionResult, and the ExecutionResult provides this. @JoviDeCroock @robrichard @benjie  any thoughts?

JoviDeCroock replied:
I think it's an anti-pattern to leave untyped bits, I can see a world where the incremental result is typed as

```ts
type PossibleIncrementalResults = 
  | UserFragment
  | TodoFragment
```

yaacovCR replied:
Just a factoid to add to this, because of inlining we are not 100% on the shape of the initial result either.

JoviDeCroock replied:
If the shape of the response, be that initial isn't predictable. Isn't that a big deviation from the typed nature of graphql?

benjie replied:
The shape of the response is as predictable as a response using `@skip` and `@include` where the value of the variables isn't factored into codegen.

JoviDeCroock replied:
Gotcha, yes, I thought I was missing something so

> because of inlining we are not 100% on the shape of the initial result either.

is just the addition of i.e. optional properties

yaacovCR replied:
c72159a45

yaacovCR committed:
address review feedback

use TIncrementalItem instead of TItems

to constrain to ReadonlyArray of TIncrementalItem

yaacovCR commented:
This PR fixes the types for Subsequent Execution Results by splitting up the generics between the deferred result data and the streamed item type, so that the constituent types can now be properly typed.

yaacovCR commented:
closing in favor of fresher #4560
